### ISSUE TITLE
[3] Imported txt files are re-indexed so the data is available

### ISSUE THREAD

**Issue by DiemBTran** [scientist-softserv/adventist-dl#117]
# Summary
Adventist is ingesting txt files that have the full text of the documents.  Some of these files will have coordinates, and some will just be text files.  The client would like the data in these txt files to be searchable as a full text search.  Do not yet consider the coordinate files.

# Acceptance Criteria

- [ ] txt files that have been imported are re-indexed so that the data is available.
- [ ] the site search (e.g. /catalog) will return results from a work that has matching search query in the associated work's txt file.

# Screenshots or Video

# Testing Instructions

## New Content

Create a plain text file (using Notepad or your text editor); in that file add some unique text.  Now create a new work and attach that file.  Save the work (wait for characterization and processing to complete).  Then go to the application's /catalog page and search for that unique text.

## Existing Content

As of 2022-10-14 this is not yet completed.

# Notes
We believe these are not working because they are not indexed.  It is possible that once they are indexed we will find another bug in the system as well.  If that is the case, create another ticket for exposed bugs.  

Confused how to fill this out? Check out the [playbook](https://playbook-staging.notch8.com/en/notch8/process/issues)

**Comment by DiemBTran** [comment]
Related MRs:

- [Adding logic to extract text from plain text files](https://gitlab.com/notch8/adventist-dl/-/merge_requests/199)
!199
- [Adding job to find non-FTS-ed text files](https://gitlab.com/notch8/adventist-dl/-/merge_requests/201)
!201

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022:

---

<time datetime="2022-10-13">Yesterday</time> in [Slack](https://assaydepot.slack.com/archives/G0313NK5NMA/p1665679243304799) I asked the following:

> Quick Guidance Requested: Full Text Search in a [Hyku](https://github.com/samvera/hyku) App For Adventist.  They have `txt` files associated with their uploaded works.  I want to index those files so that folks can search via the Hyku UI for that text.  (Not search in [Universal Viewer](http://universalviewer.io)).
> 
> My understanding from [yesterday’s Adventist project team meeting](https://docs.google.com/document/d/1AE6p5YQIECCGp7kjTF5uAS2-1TAGQBG9tfx_leQ1gDc/edit#bookmark=id.ijq1qqqewa1u) was that this may be a built-in feature in Hyku/[Hyrax](20221009T120543--the-mistimed-scroll-index_burning-wheel_campaigns_eberron_fantasy_index_mistimed-scroll_rpgs.md).  I’d love some way finding to get started on this.
> 
> For reference, here’s the [associated ticket](https://gitlab.com/notch8/adventist-dl/-/issues/147).

They are adding to the work a file\_set that has a `txt` file.

In [Hyrax::FileSetDerivativesService#extract\_full\_text](https://github.com/samvera/hyrax/blob/426575a9065a5dd3b30f458f5589a0a705ad7be2/app/services/hyrax/file_set_derivatives_service.rb#L103-L107) we conditionally extract text from the file.  This leverages `Hydra::Derivatives::FullTextExtract.create` (from v3.5.0).  Which relies on Solr’s text extraction handler.

It is unclear if they want to extract text from other files, such as any PDFs that they upload.

According to the configuration `Hyrax.config.extract_full_text?` is `true`.  Which means we could assume that a text file should be getting indexed for full text.  But since it is not, we should look further up in the `Hyrax::FileSetDerivativesService` class.

However, it appears that we are ignoring these files.  Below is the switch statement for making derivatives:

```ruby
    def create_derivatives(filename)
      case mime_type
        when *file_set.class.pdf_mime_types             then create_pdf_derivatives(filename)
        when *file_set.class.office_document_mime_types then create_office_document_derivatives(filename)
        when *file_set.class.audio_mime_types           then create_audio_derivatives(filename)
        when *file_set.class.video_mime_types           then create_video_derivatives(filename)
        when *file_set.class.image_mime_types           then create_image_derivatives(filename)
      end
    end
```

The supported `mime_types` are defined here:

```ruby
    def supported_mime_types
      file_set.class.pdf_mime_types +
        file_set.class.office_document_mime_types +
        file_set.class.audio_mime_types +
        file_set.class.video_mime_types +
        file_set.class.image_mime_types
    end
```

Which appears to indicate that we are ignoring all text files that are uploaded to the file set.  This sounds like a overlooked functionality in Hyrax; namely if I am uploading a text file I would assume that I could search on that text file.

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022:

---

We upload the file and assume the derivative service will extract the text.  I see two options:

1.  We amend the `Hyrax::DerivativeService.services`; this follows the pattern in NNP and the NewsPaper gem.
2.  We decorate the existing `Hyrax::FileSetDerivativesService` to properly handle the `txt` mime type.

The decorate pattern creates potential on-going maintenance issues; namely keeping the overrides in sync with the upstream changes.  Whereas leveraging the current configurability helps us eliminate an override.  In yet to be released [Hyrax::DerivativeService updates](https://github.com/samvera/hyrax/blob/ebc221ac3a353b150c27e2d215ea94398877b567/app/services/hyrax/derivative_service.rb#L15-L24), Hyrax is exposing (as a public API) that previously undocumented configuration point.

The proposal is to create a `Hyrax::TextFileTextExtractionService` that conforms to the defined interface of a `Hyrax::DerivativeService.services`.  We place this new service ahead of the existing service class.  And the `valid?` function would be “is this a plain text file?”  If so, grab the files text and follow the logic informed by [Hyrax::FileSetDerivativesService#extract\_full\_text](https://github.com/samvera/hyrax/blob/426575a9065a5dd3b30f458f5589a0a705ad7be2/app/services/hyrax/file_set_derivatives_service.rb#L103-L107) and subsequence function calls.

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022:

---

Note, we do not yet have the script that we need to run to update existing data.

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022:

---

Work that remains:

Create a script that finds all file sets with mime-type that matches ="text/plain*"=.

For each matching file set apply the logic of the [[https://gitlab.com/notch8/adventist-dl/blob/402a7af4d48605aa864cdb72d06dab0f3dec14ac/app/services/adventist/text_file_text_extraction_service.rb#L32-39][Adventist::TextFileTextExtractionService#create_derivative]].

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022

---

Below is the proposed script to run and test:

```ruby
FileSet.where(mime_type_ssi: 'text/plain').find_each do |file_set|
  next if file_set.extracted_text.present?

  file = file_set.original_file
  file_set.build_extracted_text.tap do |extracted_text|
    extracted_text.content = file.content
    extracted_text.mime_type = file_set.mime_type
    extracted_text.original_name = Array(file.file_name).first
  end.save
end
```

**Comment by DiemBTran** [comment]
Diem - Oct 2022

---

**This passes SoftServ QA:** See attached screenshot.

tested on:
- work with a .txt file that was created on 2021-12-07: [10-mile walk-a-thon for investment](https://sdapi.s2.adventistdigitallibrary.org/concern/journal_articles/13129204_10_mile_walk_a_thon_for_investment?locale=en)
- .txt file: [13129204 (1).txt](https://github.com/scientist-softserv/adventist-dl/files/9956201/13129204.1.txt)
- OCR search result in catalog: https://sdapi.s2.adventistdigitallibrary.org/catalog?utf8=%E2%9C%93&locale=en&search_field=all_fields&q=Sunday+morning+June+10

The .txt file has a string that says Sunday morning June 10, so I did a search for that in the catalog and found the correct work

---


![Image](https://user-images.githubusercontent.com/29311858/200434009-c054c2d1-fb5b-4ad3-a895-5e04ff14998b.png)

**Comment by DiemBTran** [comment]
Jeremy - Oct 2022:

---

I have run the script on the staging and dev site; I have **not** run this on production.

**Comment by DiemBTran** [comment]
Jeremy - Nov 2022:

---

I have run the script on production and have the following yet to review logs: 

[2022-11-03-production-index-plain-text-files-job.txt](https://github.com/scientist-softserv/adventist-dl/files/9956211/2022-11-03-production-index-plain-text-files-job.txt)

**Comment by DiemBTran** [comment]
Jeremy - Nov 2022:

---

Reviewing the logs, this is about 25% of the expected files processed. I will need to run another job to try to pick up the remaining 75%; it's possible that a move to Github from Gitlab disrupted the workers.

**Comment by jeremyf** [comment]
I have run the script again and it has picked up more files; but certainly not all of them.  My suspicion is that the workers are being shut down after awhile.  In part because the worker looks like it was restarted during that window.

**Comment by jeremyf** [comment]
I have looked at the memory setup and the workers are hitting the boundary of their memory allocation; I'm working on other approaches (namely narrower pagination and increasing memory allocation)

**Comment by KatharineV** [comment]
I tested this ticket on staging yesterday (1/17/23) by uploading a work with a txt file containing the word "supercalifragilisticexpialidocious." A search for that word doesn't return the work, despite the fact that the txt file is uploaded and has been on the site for almost a day. Perhaps something isn't working? Here's the work I tested: https://adl.s2.adventistdigitallibrary.org/concern/generic_works/d37052e6-0e13-4e8d-8216-44654ec7a29a?locale=en

**Comment by jeremyf** [comment]
There might be an inter-play between this ticket and https://github.com/scientist-softserv/adventist-dl/pull/184.  I'm investigating further.

**Comment by ShanaLMoore** [comment]
Meet with @jeremyf before starting this ticket. cc @kirkkwang @laritakr @AKFosterDev

**Comment by jeremyf** [comment]
I have, via the UI, created a work with a plain text file.  And then tested searching for a value in that file.  It did not work.

**Comment by jeremyf** [comment]
In the [./config/application.rb](https://github.com/scientist-softserv/adventist-dl/blob/d36979353238bda0feecdde26769de250e709b83/config/application.rb#L65-L70) we are setting the Hyrax::DerivativeService.services (see [Hyrax 2.9.6 implementation](https://github.com/samvera/hyrax/blob/2.9.6/app/services/hyrax/derivative_service.rb))

```ruby
# See https://gitlab.com/notch8/adventist-dl/-/issues/147
#
# By default plain text files are not processed for text extraction.  In adding
# Adventist::TextFileTextExtractionService to the beginning of the services array we are
# enabling text extraction from plain text files.
Hyrax::DerivativeService.services.unshift(Adventist::TextFileTextExtractionService)
```

What we need to do is in the UI, create work and attach a plain text file.  We want to debug and check if the [Adventist::TextFileTextExtractionService#valid?](https://github.com/scientist-softserv/adventist-dl/blob/d36979353238bda0feecdde26769de250e709b83/app/services/adventist/text_file_text_extraction_service.rb#L68-L74) method is being called and answering "true".

```ruby
def valid?
  return true if VALID_MIME_TYPES.detect do |valid_mime_type|
    # Because character encoding may be part of the mime_type.  So we want both "text/plain" and
    # "text/plain;charset=UTF-8" to be a valid type.
    valid_mime_type.start_with?(mime_type)
  end
end
```

Related to the `#valid?` is [Adventist::TextFileTextExtractionService#create\_derivatives](https://github.com/scientist-softserv/adventist-dl/blob/d36979353238bda0feecdde26769de250e709b83/app/services/adventist/text_file_text_extraction_service.rb#L55-L61).  If `#valid?` is true, we want to debug the `#create_derivatives` method.

```ruby
def create_derivatives(filename)
  self.class.assign_extracted_text(
    file_set: file_set,
    text: File.read(filename),
    original_file_name: filename
  )
end
```

Ultimately, we need the `file_set#extracted_text` to be set.  Because our FileSet indexer takes that value and adds it to SOLR (see  [Hyrax::FileSetIndexer#generate\_solr\_document](https://github.com/scientist-softserv/adventist-dl/blob/d36979353238bda0feecdde26769de250e709b83/app/indexers/hyrax/file_set_indexer.rb#L19))

```ruby
solr_doc['all_text_timv'] = Hyku.utf_8_encode(object.extracted_text.content) if object.extracted_text.present?
```

I created a work in the UI and attaching a plain text file, I found that the file\_set’s `extracted_text` was not being set.  It should be valid and done in the `Adventist::TextFileTextExtractionService`.

When I did this in the command-line, I found the file\_set (remember to switch! accounts).  I checked the file\_set.extracted\_text it was nil.  I then checked the file\_set.files.first.content, and it was the plain text file’s content (e.g. human readable words).

I ran the following outside of the request cycle.
`Adventist::TextFileTextExtractionService.assign_extracted_text(file_set: file_set, original_file_name: "words.txt", text: file_set.files.first.content)`.  I called that, which saved the file\_set.  Then I re-found that file\_set and checked the `file_set.extracted_text`, and it was set.  I rechecked in the UI, and I still couldn't find it.  So I ran `file_set.parents.first.update_index` to update the parent work's index, and then it was found.

**Comment by jeremyf** [comment]
Note the code that introduced this work and to which I really thought was working was in SHA 92c505069142cd79b752c1c67e18c382259e7058.

**Comment by kirkkwang** [comment]
tested on staging and it seems to be working for me

<img width="860" alt="image" src="https://user-images.githubusercontent.com/19597776/216457178-3b5090c0-60dc-4d77-ae64-a8f2f0131359.png">

**Comment by kirkkwang** [comment]
Could not replicate undesired behavior, we were able to upload a txt file and do a search and return the work.

<img width="927" alt="image" src="https://user-images.githubusercontent.com/19597776/216472823-8f34ba9f-156d-402c-bfaf-010cdfb72f89.png">

**Comment by KatharineV** [comment]
Confirmed this is working for a new test file I uploaded to staging today (2/3).
